fancyimpute/missforest.py

```diff
@@ -0,0 +1,249 @@
+        """
+        self.coltype_dict = None
+        self.mask_memo_dict = None
+        self.sorted_col = None
+        self.stop = False
+        self.rf_reg = RandomForestRegressor(n_estimators=n_estimators,
+                                            max_depth=max_depth,
+                                            min_samples_leaf=min_samples_leaf,
+                                            max_features=max_features,
+                                            max_samples=max_samples,
+                                            min_samples_split=min_samples_split)
+        self.rf_cla = RandomForestClassifier(n_estimators=n_estimators,
+                                            max_depth=max_depth,
+                                            min_samples_leaf=min_samples_leaf,
+                                            max_features=max_features,
+                                            max_samples=max_samples,
+                                            min_samples_split=min_samples_split)
+        self.imp_continuous_index = None
+        self.imp_categorical_index = None
+
+        Solver.__init__(self,
+            min_value=min_value,
+            max_value=max_value)
+
+    def solve(self, X, missing_mask):
+        X = check_array(X, force_all_finite=False)
+        self.sorted_col = self.sort_col(missing_mask)
+        self.coltype_dict = self._judge_type(X)
+
+        self.imp_continuous_index, self.imp_categorical_index = \
+                self.get_type_index(missing_mask, self.coltype_dict)
+
+        differ_categorical = float('inf')
+        differ_continuous = float('inf')
+
+        init_fill = X
+
+        while self.stop is False:
+
+            differ_categorical_old = differ_categorical
+            differ_continuous_old = differ_continuous
+
+            x_old_imp = init_fill
+
+            x_new_imp = []
+
+            for col in self.sorted_col:
+                tmp = []
+                if self.coltype_dict[col] is 'categorical':
+                    model = self.rf_cla
+                else:
+                    model = self.rf_reg
+
+                x_obs, y_obs, x_mis = self.split(init_fill, col, missing_mask)
+                model.fit(x_obs, y_obs)
+                y_mis = model.predict(x_mis)
+                for ele in y_mis:
+                    tmp.append(ele)
+                    x_new_imp.append(ele)
+                init_fill[:, col][missing_mask[:,col]] = tmp
+            x_new_imp = np.asarray(x_new_imp)
+
+            differ_continuous, differ_categorical = self._lose_func(x_new_imp, x_old_imp)
+            if differ_continuous >= differ_continuous_old and differ_categorical >= differ_categorical_old:
+                self.stop = True
+        return init_fill
+
+    def _lose_func(self, imp_new, imp_old):
+        """
+        Evaluation Method, mathematical concept are available at 'https://www.stu-zhouyc.com/iterForest/metrics'

```

Comment:
This website doesn't seem to exist?

Reply:
Well, thank you for your reply! It looks like scikit-learn already has MICE and missForest implementations. I should not implement these methods anymore.